packages/iTwinUI-react/src/core/ComboBox/ComboBoxMenu.tsx

```diff
@@ -1,38 +1,121 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
 import cx from 'classnames';
 import React from 'react';
 import { Menu, MenuProps } from '../Menu';
-import { useSafeContext, useMergedRefs } from '../utils';
+import { Surface } from '../Surface';
+import {
+  useSafeContext,
+  useMergedRefs,
+  useVirtualization,
+  VirtualScrollProps,
+  mergeRefs,
+  getWindow,
+} from '../utils';
 import { ComboBoxStateContext, ComboBoxRefsContext } from './helpers';
 
 type ComboBoxMenuProps = Omit<MenuProps, 'onClick'> &
   React.ComponentPropsWithoutRef<'ul'>;
 
+const VirtualComboboxMenu = ({ children, ...props }: VirtualScrollProps) => {
+  const { outerProps, innerProps, visibleChildren } = useVirtualization(props);
+
+  return (
+    <div {...outerProps}>
+      {React.cloneElement(
+        children as JSX.Element,
+        {
+          ref: mergeRefs(
+            innerProps.ref,
+            (children as React.FunctionComponentElement<HTMLElement>).ref,
+          ),
+          style: {
+            ...innerProps.style,
+            ...(children as JSX.Element).props.style,
+          },
+        },
+        visibleChildren,
+      )}
+    </div>
+  );
+};
+
 export const ComboBoxMenu = React.forwardRef(
```

Comment:
With the current API, virtualization will not be possible to use when we export subcomponents for composition.

I think we might want to move some of the things from context into props so that the user can easily do something like this:

```tsx
<ComboBox>
  <Combobox.Input />

  <ComboBox.Menu
     enableVirtualization
     length={myOwnOptions.length}
     itemRenderer={(index) => (
       <ComboBox.MenuItem>{myOwnOptions[index]?.label}</ComboBox.MenuItem>
     )}
     emptyStateMessage='Nothing to see here'
  />
</ComboBox>
```

Reply:
amm... But this can be done later, right? I dont think we have already done everything for Combobox to work with composition approach. You have even removed the story as an example.

Reply:
Yeah but it would be nice to avoid having to come back here and figure out any virtualization-specific logic that might need to be moved.

I removed the story because we haven't finalized and exported the subcomponents yet. I think only keyboard navigation doesn't fully work with composition. For everything else I was thinking about how to reduce future work.

Reply:
amm... Well I think current situation is same as with any other combobox component.
If you use it with Combobox and it provides the context, everything should work fine.
And if not, you need to expose needed props. But that is applicable to other parts too :/
So not sure, what is the problem or I misunderstood

Reply:
> If you use it with Combobox and it provides the context, everything should work fine.

Not true. With this approach you need to always pass the `options` prop into `ComboBox` root, which doesn't make sense when used with subcomponents.

I guess lets revisit this when we expose subcomponents.